### ISSUE TITLE
Save and load Input Map from within the Project Settings.

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#423]
**Describe the project you are working on:**
I constantly create small minimal projects if I want to try a specific feature/mechanic first before implementing it into my main project. Sometimes one or two a day.

**Describe the problem or limitation you are having in your project:**
Everytime I create a new project, I have to set all custom Input map actions again. In 95% these are the same every time - joypad, keyboard, mouse. It feels like a massive chore and waste of time. I'd much rather rename or delete what I don't need than create everything. 
It would also make it easy to share Input Maps.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
If I could click "load", select a config file or whatever and be done with it, instead of doing the same process over and over again, I would be a much more zen person.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![Load-save-inputmap](https://user-images.githubusercontent.com/47016402/73499940-c7ea6d80-43c1-11ea-86d3-380df0deec20.png)

**Describe implementation detail for your proposal (in code), if possible:**
Clicking on "Save" would open a prompt asking to select a place and a name to save the simple text config file. (res:// by default)
Clicking on "Load" would open a prompt to browse to the place to select the Input Map config file. (res:// by default)

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It would be used a lot by anyone creating small test projects, teachers, people testing things in the engine ...

**Is there a reason why this should be core and not an add-on in the asset library?:**
I think this should be build into the engine as it would speed up workflow quite a lot.

**Comment by Calinou** [comment]
Wouldn't it make more sense to allow saving and loading input map files from anywhere on the filesystem, not just `res://`?

**Comment by golddotasksquestions** [comment]
> Wouldn't it make more sense to allow saving and loading input map files from anywhere on the filesystem, not just res://?
Yes absolutely, that's what I meant with "select a place". res:// just being the default.

**Comment by FeatherAntennae** [comment]
I'm adding a menu and an option to restore default here https://github.com/godotengine/godot/pull/31815

It could help to put the option to import / export in there.

**Comment by KoBeWi** [comment]
InputMap is stored in your project, so you just need a "template" project.godot with InputMap configured and just copy it when you want to create a new project.

**Comment by FeatherAntennae** [comment]
@KoBeWi A lot of things are stored this way, importing and exporting the InputMap would allow to import and export _only_ the InputMap. I can see this be useful.

**Comment by golddotasksquestions** [comment]
@KoBeWi  That's a good tip if you'd want to copy more than just the Input map. Thx!

**Comment by haennes** [comment]
I think this should be implemented.
It would make changeable InputMaps much easier.

**Comment by pycbouh** [comment]
@haennes Please use the thumb up reaction on the first message to show your support. Use comments only if there is something you want to add or ask on the matter.

**Comment by haennes** [comment]
> @haennes Please use the thumb up reaction on the first message to show your support. Use comments only if there is something you want to add or ask on the matter.

just changed my comment
can you unmark it as off-topic
i will delete this after you have read this

**Comment by TheOrioli** [comment]
The way Unity handles this is with an [InputActionAsset ](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.1/api/UnityEngine.InputSystem.InputActionAsset.html). 

The added benefit is you can ship the game with multiple control schemes and switch them out on the fly through an options menu if you don't want to provide full control customization.

**Comment by KoBeWi** [comment]
I created an asset that can save your remap to a file and allows for easy editing of controls:
https://github.com/KoBeWi/Godot-Input-Remap
It's designed for easy controls customization in games.

**Comment by Pinkyyyy** [comment]
Was just googling trying to figure out if importing an input map exists, and it seems like it doesn't still, any plans for this in the future?

**Comment by Calinou** [comment]
> Was just googling trying to figure out if importing an input map exists, and it seems like it doesn't still, any plans for this in the future?

4.0 is in feature freeze, and there is no ETA for features being implemented. Also, we still haven't reached a consensus whether this should be editor-only functionality, or something that can also be used in exported projects. Another implementation detail is that whether the saving/loading process should affect built-in actions such as `ui_accept`.

**Comment by Pinkyyyy** [comment]
> 4.0 is in feature freeze, and there is no ETA for features being implemented. Also, there still isn't a clear consensus onwhether this should be editor-only functionality, or something that can also be used in exported projects.

Hopefully it's there for 4.1, seems like a really useful thing to have.
> Another question we haven't decided upon yet is whether the saving/loading process should affect built-in actions such as `ui_accept`.

Personally since 4.0 seems to hide built-in actions they shouldn't be affected by an imported input map.

I guess a way to do this without having to import an input map or rather have to create one from scratch every time you make a FPS game for example is by just using template projects.

**Comment by Shadowblitz16** [comment]
> > Was just googling trying to figure out if importing an input map exists, and it seems like it doesn't still, any plans for this in the future?
> 
> 4.0 is in feature freeze, and there is no ETA for features being implemented. Also, there still isn't a clear consensus on whether this should be editor-only functionality, or something that can also be used in exported projects.
> 
> Another question we haven't decided upon yet is whether the saving/loading process should affect built-in actions such as `ui_accept`.

It should NOT be editor only.

I honestly think ui actions should be bindable to input map resources.
It would be nice to treat separate ui with separate actions and if the multi ui focus gets implemented it would be extra useful

**Comment by Catastrphic** [comment]
a nice way to implement it is to have an option to include built in actions and to be able to do this from a script.

**Comment by EladKarni** [comment]
I'm not sure if this is still an active proposal, but I ended up making a plugin for this. I'm hoping it'll help with the proposal so we can have this feature added to the main engine 😄 

https://github.com/EladKarni/InputMapperPresetLoader

Should I make a new proposal since this one is from 2020 or should I do something else to try and promote this feature?

**Comment by golddotasksquestions** [comment]
Instead of another proposal for the same thing, you can submit a Pull Request which would solve this issue @EladKarni

**Comment by ZycaR** [comment]
@golddotasksquestions - I'm new here and I accidentally created similar proposal. (should be closed now) Can you please revisit my proposal if it is aligned with your needs?
BTW: working on the solution into curent code v4.2

**Comment by ZycaR** [comment]
Pull request is kind of rerady to review. Is it possible to give feedback if it is ok, or there is additional work?

The code change adds 'three dot' menu with import/export:
![284427615-b2963c78-5d76-410b-9d22-c34529ffb066](https://github.com/godotengine/godot-proposals/assets/11602729/b5633715-841b-4a65-82f2-d3207c3370e4)
